src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -41,7 +41,31 @@ private static void RootReflectionSerializerDependencies()
             }
 
             [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-            static JsonTypeInfo CreateJsonTypeInfo(Type type, JsonSerializerOptions options) => new JsonTypeInfo(type, options);
+            static JsonTypeInfo CreateJsonTypeInfo(Type type, JsonSerializerOptions options)
+            {
+                JsonTypeInfo.ValidateType(type, null, null, options);
+
+                MethodInfo methodInfo = typeof(JsonSerializerOptions).GetMethod(nameof(CreateReflectionJsonTypeInfo), BindingFlags.NonPublic | BindingFlags.Instance)!;
+#if NETCOREAPP
+                return (JsonTypeInfo)methodInfo.MakeGenericMethod(type).Invoke(options, BindingFlags.NonPublic | BindingFlags.DoNotWrapExceptions, null, null, null)!;
+#else
+                try
+                {
+                    return (JsonTypeInfo)methodInfo.MakeGenericMethod(type).Invoke(options, null)!;
+                }
+                catch (TargetInvocationException ex)
+                {
+                    throw ex.InnerException!;
+                }
+#endif
+            }
+        }
+
+        [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+        private JsonTypeInfo<T> CreateReflectionJsonTypeInfo<T>()
+        {
+            // We do not use Activator.CreateInstance because it will wrap exception if constructor throws it
```

Comment:
We use Activator.CreateInstance pretty liberally when constructing converters. Wouldn't this be a concern there as well?

Reply:
More importantly, if this is throwing, isn't this an indication of an internal bug? Do we need to care about this exception being surfaced appropriately to users?

Reply:
Actually that's incorrect comment, I forgot to remove it. I used it here because it makes debugging easier and also I don't need to pass options to Invoke when using instance member (one allocation less)

Reply:
For throwin indicating internal bug:
- not always: JsonTypeInfo can be constructed using source gen APIs and if there is not enough information given in there Configure will throw an exception - we currently have a set of tests checking if we throw an appropriate exception and the construction through reflection causes them to be wrapped which makes the experience weird

Reply:
If the constructor is capable of surfacing exceptions relevant to the user, we might be able to work around the issue by making the constructor itself trivial (i.e. just create the instance with the T type) and then follow up with a `Configure()` method on the created instance. I believe we already follow this pattern in JsonPropertyInfo.

Reply:
I've just quickly checked which tests are failing when unwrapping is commented out and seeing following types of failures we check for (84 tests failures total in System.Text.Json.Tests alone):
- property/element/key type is not valid for serialization - this is happening when converter is determined and converter is required for constructing JsonTypeInfo
- while determining properties: invalid naming policies etc. - this could potentially be delayed although it cannot be moved to `Configure` - the latest we can move that to is when user first time accesses properties (we will have to do that for source gen already)
- invalid converter (i.e. different converter type that property type) - again this should be done during construction
- extension property is of wrong type - this could also potentially be slightly delayed (to first time accessing properties)
- invalid property type
- couple of other sanity errors (i.e. combining number handling with custom converter for numbers)
I'll stop checking here.

Given above I don't think it can really be delayed. Also note that this is happening only during construction of JsonTypeInfo which is a one time operation